plugin.video.ted.talks/resources/lib/ted_talks.py

```diff
@@ -0,0 +1,339 @@
+import xbmcvfs
+
+from . import menu_util
+from . import plugin
+from . import settings
+from . import ted_talks_scraper
+from .model.fetcher import Fetcher
+from .model.rss_scraper import NewTalksRss
+from .model.search_scraper import Search
+from .model.speakers_scraper import Speakers
+from .model.topics_scraper import Topics
+
+class UI:
+
+    def __init__(self, fetcher, ted_talks):
+        self.fetcher = fetcher
+        self.ted_talks = ted_talks
+
+    def endofdirectory(self, sortMethod='title', updateListing=False):
+        # set sortmethod to something xbmc can use
+        if sortMethod == 'title':
+            sortMethod = xbmcplugin.SORT_METHOD_LABEL
+        elif sortMethod == 'date':
+            sortMethod = xbmcplugin.SORT_METHOD_DATE
+        elif sortMethod == 'none':
+            sortMethod = xbmcplugin.SORT_METHOD_NONE
+
+        # Sort methods are required in library mode.
+        xbmcplugin.addSortMethod(int(sys.argv[1]), sortMethod)
+        # let xbmc know the script is done adding items to the list.
+        xbmcplugin.endOfDirectory(handle=int(sys.argv[1]), updateListing=updateListing)
+
+    def addItem(self, title, mode, url=None, img='', args={}, video_info={}, isFolder=True, total_items=0):
+        # Create action url
+        args['mode'] = mode
+        if url:
+            args['url'] = url
+
+        args = [k + '=' + urllib.parse.quote_plus(v.encode('ascii', 'ignore')) for k, v in args.items()]
+        action_url = sys.argv[0] + '?' + "&".join(args)
+
+        li = xbmcgui.ListItem(label=title)
+        li.setArt({'icon': img, 'thumb': img})
+        video_info = dict((k, v) for k, v in video_info.items() if k in ['date', 'plot', 'mediatype'])
+        li.setInfo('video', video_info)
+        if 'duration' in video_info:
+            # To set with second granularity must do this rather than via setInfo
+            li.addStreamInfo('video', {'duration' : video_info['duration']})
+        if not isFolder:
+            li.setProperty("IsPlayable", "true")  # let xbmc know this can be played, unlike a folder.
+            context_menu = menu_util.create_context_menu(getLS=plugin.getLS)
+            li.addContextMenuItems(context_menu, replaceItems=False)
+        else:
+            li.addContextMenuItems([], replaceItems=False)
+        xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=action_url, listitem=li, isFolder=isFolder, totalItems=total_items)
+
+    def playVideo(self, url, icon):
+
+        plugin.report('%s = %s' % ('url', str(url)), level='debug')
+
+        subs_language = settings.get_subtitle_languages()
+        playlist, title, subs, info_labels = self.ted_talks.get_video_details(url=url, subs_language=subs_language)
+
+        playlist_file = os.path.join(xbmcvfs.translatePath("special://temp/"), 'ted_talks.m3u8')
+        with open(playlist_file, 'w', encoding='utf-8') as fh:
+            fh.write(playlist)
+
+        li = xbmcgui.ListItem(title, path=playlist_file)
```

Comment:
consider to define `offscreen=True` in all your listitems. See https://codedocs.xyz/xbmc/xbmc/group__python__xbmcgui__listitem.html#ga15bcb930a4f91dd97f8ed7c633a20618